### ISSUE TITLE
Revision model rename breaks Page revision foreign key on SQLite

### ISSUE THREAD

**Issue by chosak** [wagtail/wagtail#8635]
### Issue Summary

#8441 renamed the PageRevision model to Revision, which included a migration with [a `RenameModel` step](https://github.com/wagtail/wagtail/blob/1f43d8ef51e455b92e42447fdc190d5ec83ec53c/wagtail/migrations/0070_rename_pagerevision_revision.py#L15-L18).

On my local machine, running against SQLite in the default configuration, this renamed the table but didn't update the foreign key from the Page model. Looking at the SQL for the migration starts with:

```
% ./manage.py sqlmigrate wagtailcore 0070
BEGIN;
--
-- Rename model PageRevision to Revision
--
ALTER TABLE "wagtailcore_pagerevision" RENAME TO "wagtailcore_revision";
...
```

But if I then check the `live_revision_id` foreign key on the Page model in SQLite, it hasn't been updated, and still points to the now-renamed `wagtailcore_pagerevision` table.

```
% sqlite3 db.sqlite3 
SQLite version 3.32.3 2020-06-18 14:16:19
Enter ".help" for usage hints.
sqlite> PRAGMA foreign_key_list('wagtailcore_page');
...
3|0|wagtailcore_pagerevision|live_revision_id|id|NO ACTION|NO ACTION|NONE
...
```

It looks like I'm getting hit by the `ALTER TABLE RENAME` issue described in the SQLite docs [here](https://www.sqlite.org/draft/lang_altertable.html#alter_table_rename):

> Beginning with version 3.26.0, FOREIGN KEY constraints are always converted when a table is renamed, unless the [PRAGMA legacy_alter_table=ON](https://www.sqlite.org/draft/pragma.html#pragma_legacy_alter_table) setting is engaged.

My `PRAGMA`s are defined thusly (the default values, apparently, for this version of SQLite on MacOS):

```
sqlite> PRAGMA legacy_alter_table;
1
sqlite> PRAGMA foreign_keys;
0
```

I note [this commit](https://github.com/django/django/commit/063cf98d3a6839f40c423cbd845def429c5cf0ce) that just went into Django (the dev version for 4.1) that explicitly disables `legacy_alter_table`; I wonder if I am hitting some edge case that this would fix.

Wagtail seems to have only [one other instance of `RenameModel`](https://github.com/wagtail/wagtail/blob/716bf92c2dc2da2aca5e8f5aa6768b5b087cd4b0/wagtail/contrib/search_promotions/migrations/0001_initial.py#L65) for the SearchPromotion model, but I don't think we have any foreign keys pointing to that. So this might be the first time this combination has hit Wagtail.

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. `cd myproject`
3. `./manage.py migrate`
4. `./manage.py createsuperuser` and create an admin user.
5. `./manage.py runserver`
6. Vist http://localhost:8000/admin/pages/add/home/homepage/3/ to create a new page (you'll be asked to log in first)
7. Fill in the title field with something like "Test". Click "Save Draft". You'll get an error: "OperationalError at /admin/pages/add/home/homepage/3/, no such table: main.wagtailcore_pagerevision"

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

- Mac OS version: Big Sur 11.6.5
- SQLite3 version: 3.32.3 2020-06-18 14:16:19 02c344aceaea0d177dd42e62c8541e3cab4a26c757ba33b3a31a43ccc7d4aapl
-   Python version: 3.10.4
-   Django version: 4.0.5, also happens on 3.2
-   Wagtail version: main (1f43d8ef51e455b92e42447fdc190d5ec83ec53c)

**Comment by thibaudcolas** [comment]
Just to report I think I have the same issue here. Also hit when creating a new page.

I get the same output as @chosak for `PRAGMA foreign_key_list('wagtailcore_page')`, `PRAGMA legacy_alter_table`, `PRAGMA foreign_keys`

This is on macOS 12.3 Monterey, SQLite version 3.37.0 2021-12-09 01:34:53, same Wagtail commit 1f43d8ef51.

**Comment by chosak** [comment]
Looking into this a bit more, this definitely seems like a Django issue and nothing specific to Wagtail. That being said, we might need to find a workaround if we want to release this table rename given current Django logic.

I dug into the Django history around these SQLite pragmas (which is quite complicated!) and I think this is the same issue as Django [ticket 31765](https://code.djangoproject.com/ticket/31765), which was fixed in https://github.com/django/django/commit/063cf98d3a6839f40c423cbd845def429c5cf0ce (pending Django release 4.1). If I backport this change to Django 4.0, then my local combination of MacOS/SQLite works properly to repoint the foreign key, and Wagtail page creation is fixed.

I commented at https://code.djangoproject.com/ticket/31765#comment:20 to see if it might be possible to get that commit backported to current versions of Django. Otherwise I wonder if we might try adding a similar fix to our problematic migration (something like `PRAGMA legacy_alter_table = OFF`). Obviously it would be nicer not to have to do that.

**Comment by laymonage** [comment]
Thanks for the report!

I think it's worth noting that the SQLite installed on the system is not always the same as the SQLite bundled in Python. In my case:

<img width="744" alt="image" src="https://user-images.githubusercontent.com/6379424/172661854-42e942b3-29ed-4669-b86d-4e1ace58e34d.png">

As the `legacy_alter_table` is `0` on my Python's SQLite, I did not encounter the issue.

My environment:
- macOS Monterey 12.4 on MacBook Pro M1 Pro
- Python 3.10.2 installed from [Python.org](https://python.org)
- SQLite version: 3.35.5 (Python), 3.37.0 (System)
- Django version: 4.0.5
- Wagtail version: `main`

**Comment by thibaudcolas** [comment]
👍 here are the results of the above for me:

```console
$ python
Python 3.10.0  […]
$ import sqlite3
$ conn = sqlite3.connect(':memory:')
$ cursor = conn.cursor()
$ cursor.execute('PRAGMA legacy_alter_table').fetchone()
(1,)
$ cursor.execute('PRAGMA foreign_keys').fetchone()
(0,)
$ sqlite3.sqlite_version
'3.37.0'
```

**Comment by laymonage** [comment]
Hey all, I made a fix in #8708 by executing `PRAGMA legacy_alter_table = OFF` before running the migration operations. You'll probably have to rebuild your SQLite database though, as rolling back the migration might not revert all the changes. Let me know if this is still an issue. Thanks!

**Comment by Scotchester** [comment]
Hey @laymonage, I'm just getting around to trying this again and it's still not working for me in my bakerydemo repo with Wagtail's `main` branch installed. Your fix in #8708 doesn't seem to be having an effect. I get the same error when hitting the migration:

```
Applying wagtailcore.0070_rename_pagerevision_revision...Traceback (most recent call last):
  File "./manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/core/management/base.py", line 414, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/core/management/base.py", line 460, in execute
    output = self.handle(*args, **options)
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/core/management/base.py", line 98, in wrapped
    res = handle_func(*args, **kwargs)
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 290, in handle
    post_migrate_state = executor.migrate(
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 131, in migrate
    state = self._migrate_all_forwards(
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 163, in _migrate_all_forwards
    state = self.apply_migration(
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 248, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/db/migrations/migration.py", line 131, in apply
    operation.database_forwards(
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/db/migrations/operations/models.py", line 359, in database_forwards
    schema_editor.alter_db_table(
  File "/Users/scott/Projects/bakerydemo/.venv/lib/python3.8/site-packages/django/db/backends/sqlite3/schema.py", line 107, in alter_db_table
    raise NotSupportedError(
django.db.utils.NotSupportedError: Renaming the 'wagtailcore_pagerevision' table while in a transaction is not supported on SQLite < 3.26 because it would break referential integrity. Try adding `atomic = False` to the Migration class.
```

And here's my Python console output:

```
~/Projects/bakerydemo  main ✗ ⚑                                                      [14:14:53] 
▶ python
Python 3.8.2 (default, Apr 19 2020, 23:00:02) 
[Clang 11.0.3 (clang-1103.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> sqlite3.sqlite_version
'3.28.0'
>>> conn = sqlite3.connect(':memory:')
>>> cursor = conn.cursor
>>> cursor = conn.cursor()
>>> cursor.execute('PRAGMA legacy_alter_table').fetchone()
(1,)
>>> cursor.execute('PRAGMA foreign_keys').fetchone()
(0,)
```

FWIW, things are working fine in https://github.com/wagtail/gitpod-wagtail-develop where the sqlite3 version Python uses is 3.31.0.

I'll try updating my Python version to get a newer bundled SQLite and see if that fixes it, but even if it does, I don't think we want to tell people they have to upgrade from a still-supported version of Python, do we?

**Comment by Scotchester** [comment]
Update: I figured out what was going on in my environment.

I use pyenv to install the various versions of Python I want to try. When it builds a Python from source, it brings in build dependencies from a couple potential places. Though I had installed a newer SQLite (3.39.1) via Homebrew, as recommended by pyenv docs, it was still pulling the macOS-built-in SQLite (at 3.28) when building new installs of Python. Apparently I needed to be setting some environment variables to get pyenv to find the newer SQLite installed by Homebrew.

Once I got Python built with that newer SQLite, I could finally run the migration in question.

This indicates to me that we're not out of the woods, if we intend to still support older versions of SQLite. Cross-referencing my related conversation with @gasman on #8757, do we need to determine a specific minimum version we're going to support? And if it's at least 3.28 or older (not sure where between 3.28 and 3.31 this got fixed), which Matt indicated in that thread should be the case, I think we have more fixing to do on this bug. Going to reopen this issue on that basis.

At the very least, we should probably have an upgrade consideration that users may need to upgrade SQLite if they run into this issue when trying to upgrade to Wagtail 4.

**Comment by gasman** [comment]
Not having much luck replicating this issue, on a Mac OS setup very similar to the original report (but returning legacy_alter_table = 0 when testing as per https://github.com/wagtail/wagtail/issues/8635#issuecomment-1150099316). Even upgrading to a Wagtail commit with #8708 applied, but changing `PRAGMA legacy_alter_table = OFF` to `ON`, doesn't fail.

Will have a go at testing against an old Vagrant image as per https://github.com/wagtail/wagtail/issues/7798#issuecomment-1021544265.

**Comment by gasman** [comment]
Have tested the current main branch with the ubuntu/xenial setup described in https://github.com/wagtail/wagtail/issues/7798#issuecomment-1021544265 . This gives a sqlite 3.11.0 installation, pre-dating `PRAGMA legacy_alter_table` - i.e. the legacy behaviour is always in force.

Initially the wagtailcore.0070 migration fails with:
```
django.db.utils.NotSupportedError: Renaming the 'wagtailcore_pagerevision' table while in a transaction is not supported on SQLite < 3.26 because it would break referential integrity. Try adding `atomic = False` to the Migration class.
```
Adding `atomic = False` to the Migration class as suggested allows the migration to complete, and subsequently creating a page / saving as draft works correctly.

Any reason not to make this change? I'll open a PR so we can at least see if it breaks anything on our CI runs...

**Comment by chosak** [comment]
If it's helpful to have another data point, I can confirm that both main and #8965 work on my SQLite 3.32.3 configuration [from above](https://github.com/wagtail/wagtail/issues/8635#issue-1262447443).

**Comment by gasman** [comment]
(Hopefully) fixed in #8965.
